feat: allow defining a git service from defaults.ini (#694)
oracle/macaron: docs/source/pages/using.rst

If you have a local repository that you want to analyze, Macaron also supports running the analysis against a local repository.

''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
Analyzing a repository whose git service is not supported by Macaron
''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''

If the repository remote URL is from an unknown git service (see :ref:`Git Services <supported_git_services>` for a list of supported git services in Macaron), Macaron won't recognize it when analyzing the repository.

You would need to tell Macaron about that git service through the ``defaults.ini`` config.

> For example, let's say you want to analyze a repository hosted at ``https://git.example.com/foo/target``. First, you need to create a ``defaults.ini`` file in the current workspace with the following content:

Reviewer: I think the word `current workspace` here may be confusing. It is probably better to say `current working directory`.

Author: Fixed in https://github.com/oracle/macaron/pull/694/commits/52e50f7d6799676ff8095cbc3c5082a158c2da8a

> For example, let's say you want to analyze the Bitbucket repository at ``https://bitbucket.org/snakeyaml/snakeyaml``. First, you need to create a ``defaults.ini`` file in the current workspace with the following content:

Reviewer 2: I'm not sure using  `bitbucket.org` as the example is a good idea, since this is a well-known public git service that in principle we should support at some point, rendering this example meaningless (and the same goes for any other real public git service one might choose). Using a nonexistent hostname as the example (e.g. `git.example.com`) would probably be better, at the cost of making the example not concretely executable.

Author: I agree. Using `bitbucket.org` is not the best way to demonstrate an unknown generic git service.

.. code-block:: ini

  [git_service.local_repo]
  hostname = git.example.com

> In which ``hostname`` contains the domain of the git service URL. In this example it's ``git.example.com``.

Reviewer: ```suggestion
In which ``hostname`` contains the hostname of the git service URL. In this example, it is ``git.example.com``.
```

The term "domain" used here may not be the most accurate. We had something similar to this before. See #416.

Author: Fixed in https://github.com/oracle/macaron/pull/694/commits/ace778c0b7ee12da54461140ddee63ff2ebb1f13 and https://github.com/oracle/macaron/pull/694/commits/a96ab2bbd8d6ff29f32e30b77619e8326cc31fa9

.. note::

>   This ``defaults.ini`` section must only be used for analyzing a locally cloned repository. If the domain name has already been supported in other services, it doesn't need to be defined again here.

Reviewer: A bit nitpicking, but I think this is more general:
```suggestion
  This ``defaults.ini`` section must only be used for analyzing a repository on the local file system. If the domain name has already been supported in other services, it doesn't need to be defined again here.
```
A local repository may just have been created on the local machine and has not been pushed to any remote.

Author: Fixed in https://github.com/oracle/macaron/pull/694/commits/ace778c0b7ee12da54461140ddee63ff2ebb1f13

Assume that the dir tree at the current workspace has the following structure:

.. code-block:: shell

  boo
  ├── foo
  │   └── target

We can run Macaron against the local repository at ``target`` by using this command:

.. code-block:: shell

  ./run_macaron.sh --local-repos-path ./boo/foo --defaults-path ./defaults.ini analyze -rp target <rest_of_args>

With ``rest_of_args`` being the arguments to the ``analyze`` command (e.g. ``-b``, ``-d`` or ``--skip-deps`` similar to two previous examples).

> The ``-lr`` flag tells Macaron to look into ``./boo/foo`` for local repositories. For more information, please see :ref:`Command Line Usage <cli-usage>`.

Reviewer: The example command uses `--local-repos-path` but `-lr` is mentioned instead here.
It is probably best to mention both the long and short versions.
```suggestion
The ``--local-repos-path/-lr`` flag tells Macaron to look into ``./boo/foo`` for local repositories. For more information, please see :ref:`Command Line Usage <cli-usage>`.
```

Author: Fixed in https://github.com/oracle/macaron/pull/694/commits/7c360b00111c1de6c824b8f90598c9c2551732aa

.. note:: If ``-lr`` is not provided, Macaron will looks inside ``<current_working_directory>/output/git_repos/local_repos/`` whenever you provide a local path to ``-rp``.

'''''''''''''''''''''''''''''''''''''''''''''''''''''''
Analyzing a local repository with supported git service
'''''''''''''''''''''''''''''''''''''''''''''''''''''''

> If the local repository you want to analyze has remote origin hosted on a supported git service, you can run the analysis directly without having to prepare ``defaults.ini`` as above.

Reviewer 3: ```suggestion
If the local repository you want to analyze has a remote origin hosted on a supported git service, you can run the analysis directly without having to prepare ``defaults.ini`` as above.
```

Author: Fixed in https://github.com/oracle/macaron/pull/694/commits/559041db4f8db569f9098ab6df056c5b3446d999

.. code-block:: shell

  ./run_macaron.sh --local-repos-path ./boo/foo analyze -rp target <rest_of_args>

With ``rest_of_args`` being the arguments to the ``analyze`` command (e.g. ``-b``, ``-d`` or ``--skip-deps`` similar to two previous examples).

> The ``-lr`` flag tells Macaron to look into ``path/to/boo/foo`` for local repositories. For more information, please see :ref:`Command Line Usage <cli-usage>`.

Reviewer 2: `path/to/boo/foo` -> `./boo/foo` to be consistent with the change above.

Author: Fixed in https://github.com/oracle/macaron/pull/694/commits/a05ef37bd7f4509e6d36378eae137153eb423cbe
